File: cpp/01-commands/doc/Banking_Kata/AccountStatementPrintTest.cpp
#include "pch.h"
#include "CppUnitTest.h"
#include "BankingTransactionDTO.h"
#include "AccountStatementPrint.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace BankingKata
{
	TEST_CLASS(AccountStatementPrintTest)
	{
	public:
		
		TEST_METHOD(WhenFirstDepositMade_DateAmountBalanceArePrinted)
		{
			std::vector<BankingTransactionDTO> statement;
            statement.push_back(BankingTransactionDTO(system_clock::now(), +500, 500));
			auto printedStatement = PrintStatement(statement);
			auto expectedStatement = std::format(
				"Date Amount Balance\n"
				"{} +500 500\n", 
				PrintDate(system_clock::now()));
			Assert::AreEqual(expectedStatement, printedStatement);
         }

		TEST_METHOD(WhenDepositAndWithdrawalMade_DateAmountBalanceArePrinted)
		{
            std::vector<BankingTransactionDTO> statement;
            statement.push_back(BankingTransactionDTO(system_clock::now(), +500, 500));
            statement.push_back(BankingTransactionDTO(system_clock::now(), -100, 400));
            auto printedStatement = PrintStatement(statement);
			auto expectedStatement = std::format(
				"Date Amount Balance\n"
				"{0} +500 500\n"
				"{0} -100 400\n",
				PrintDate(system_clock::now()));
			Assert::AreEqual(expectedStatement, printedStatement);
         }

	};
}

File: dotnet/01-commands/doc/Banking_Kata/AccountStatementPrintTest.cs
namespace Banking_Kata
{
    public class AccountStatementPrintTest
    {

        [Fact]
        public void WhenAPrintStatementIsAskedAfterTheFirstDeposit_DateAmountAndBalanceArePrinted()
        {
            //Arrange 
            List<BankingTransactionDTO> statement = new List<BankingTransactionDTO>();
            string expectedOutput = "Date Amount Balance\n" +
                                    $"{DateTime.Today.ToString("d")} +500 500\n";
            
            //Act
            statement.Add(new BankingTransactionDTO(DateTime.Today, 500, 500));


            //Assert
            Assert.Equal(expectedOutput, AccountStatementPrint.PrintStatement(statement));
        }

        [Fact]
        public void WhenAPrintStatementIsAskedAfterADepositAndAWithdraw_DateAmoutAndBalanceArePrinted()
        {
            //Arrange 
            List<BankingTransactionDTO> statement = new List<BankingTransactionDTO>();
            string expectedOutput = "Date Amount Balance\n" +
                                    $"{DateTime.Today.ToString("d")} +500 500\n" +
                                    $"{DateTime.Today.ToString("d")} -100 400\n";

            //Act
            statement.Add(new BankingTransactionDTO(DateTime.Today, 500, 500));
            statement.Add(new BankingTransactionDTO(DateTime.Today, -100, 400));

            //Assert
            Assert.Equal(expectedOutput, AccountStatementPrint.PrintStatement(statement));
        }
    }
}